lib/server/services/qualityControlFlag/QcFlagService.js

```diff
@@ -314,6 +316,117 @@ class QcFlagService {
         return summary;
     }
 
+    /**
+     * Set list of DPL detectors which contribute to GAQ (global aggregated quality)
+     * for given runs and data pass
+     *
+     * @param {object} scope of definition
+     * @param {number} scope.dataPassId data pass id
+     * @param {number[]} scope.runNumbers list of unique run numbers
+     * @param {number[]} scope.dplDetectorIds list of unique DPL detector ids
+     * @return {Promise<void>} promise
+     */
+    setGaqDetectors({ dataPassId, runNumbers, dplDetectorIds }) {
+        return dataSource.transaction(async () => {
+            await getOneDataPassOrFail({ id: dataPassId });
+            await this._validateDataPassAndRunsAssociation(dataPassId, runNumbers);
+            await this._validateRunsAndDplDetectorsAssociations(runNumbers, dplDetectorIds);
+
+            await GaqDetectorRepository.removeAll({
+                where: {
+                    dataPassId,
+                    runNumber: { [Op.in]: runNumbers },
+                },
+            });
+            const gaqEntries = runNumbers
+                .flatMap((runNumber) => dplDetectorIds
+                    .map((dplDetectorId) => ({ dataPassId, runNumber, dplDetectorId })));
+            const createdEntries = await GaqDetectorRepository.insertAll(gaqEntries);
+            return createdEntries.map(gaqDetectorAdapter.toEntity);
+        });
+    }
+
+    /**
+     * Validate association between data pass and runs exists
+     *
+     * @param {number} dataPassId id of a data pass
+     * @param {number[]} runNumbers list of run numbers
+     * @return {Promise<void>} promise resolved once associatios are validated
+     * @throws {BadParameterError} if some association does not exist
+     */
+    async _validateDataPassAndRunsAssociation(dataPassId, runNumbers) {
+        const validRunNumbers = await RunRepository.findAll({
+            attributes: ['runNumber'],
+            where: { runNumber: { [Op.in]: runNumbers } },
+            include: [{ association: 'dataPass', where: { id: dataPassId }, required: false }],
+        });
+        if (validRunNumbers.length !== runNumbers.length) {
+            const validRunNumbersSet = new Set(validRunNumbers.map(({ runNumber }) => runNumber));
+            const invalidRunNumbers = runNumbers.filter((runNumber) => !validRunNumbersSet.has(runNumber));
+            const errorMessage = `No runs for run numbers: (${invalidRunNumbers})`;
+            throw new BadParameterError(errorMessage);
+        }
+        const runsNotAssociatedToDataPass = validRunNumbers
+            .filter(({ dataPass }) => dataPass.length === 0)
+            .map(({ runNumber }) => runNumber);
+        if (runsNotAssociatedToDataPass.length > 0) {
+            const errorMessage = `No association between data pass with id ${dataPassId} and following runs: ${runsNotAssociatedToDataPass}`;
+            throw new BadParameterError(errorMessage);
+        }
+    }
```

Comment:
I think there is no need for such complexity. You should simply check that if you count runs with this id(s) associated to data pass with this id, you're good (maybe you can even create a model on the association table? Not sure it's worth it)
As mentioned before, you update GAQ runs per runs, so this is easy 

This should probably go away from QC flags too

Reply:
https://github.com/AliceO2Group/Bookkeeping/pull/1621#discussion_r1673619269

Reply:
simplified